tq/transfer_queue.go

```diff
@@ -165,12 +196,266 @@ func (q *TransferQueue) Add(t Transferable) {
 		q.transferables[t.Oid()] = t
 		q.trMutex.Unlock()
 	} else {
-		tracerx.Printf("already transferring %q, skipping duplicate", t)
 		q.trMutex.Unlock()
+
+		tracerx.Printf("already transferring %q, skipping duplicate", t.Oid())
 		return
 	}
 
-	q.batcher.Add(t)
+	q.incoming <- t
+}
+
+// collectBatches collects batches in a loop, prioritizing failed items from the
+// previous before adding new items. The process works as follows:
+//
+//   1. Create a new batch, of size `q.batchSize`, and containing no items
+//   2. While the batch contains less items than `q.batchSize` AND the channel
+//      is open, read one item from the `q.incoming` channel.
+//      a. If the read was a channel close, go to step 4.
+//      b. If the read was a Transferable item, go to step 3.
+//   3. Append the item to the batch.
+//   4. Sort the batch by descending object size, make a batch API call, send
+//      the items to the `*adapterBase`.
+//   5. Process the worker results, incrementing and appending retries if
+//      possible.
+//   6. If the `q.incoming` channel is open, go to step 2.
+//   7. If the next batch is empty AND the `q.incoming` channel is closed,
+//      terminate immediately.
+//
+// collectBatches runs in its own goroutine.
+func (q *TransferQueue) collectBatches() {
+	var closing bool
+	batch := q.makeBatch()
+
+	for {
+		for !closing && !batch.IsFull(q.batchSize) {
+			t, ok := <-q.incoming
+			if !ok {
+				closing = true
+				break
+			}
+
+			batch = append(batch, t)
+		}
+
+		// Before enqueuing the next batch, sort it's items by size
+		// (largest first) so that workers can process the largest
+		// transfers first.
+		sort.Sort(sort.Reverse(batch))
```

Comment:
OK I see, so it's saving you from the case where one massive file is the absolute last thing off the queue when everything else is already finished. I'm not sure if in practice it's likely to make much of difference but I understand where you're coming from now :)